pkg/oauthflow/device.go

```diff
@@ -0,0 +1,146 @@
+/*
+Copyright © 2021 The Sigstore Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package oauthflow
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"golang.org/x/oauth2"
+)
+
+type deviceResp struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationUri         string `json:"verification_uri"`
+	VerificationUriComplete string `json:"verification_uri_complete"`
+	Interval                int    `json:"interval"`
+}
+
+type tokenResp struct {
+	IDToken string `json:"id_token"`
+	Error   string `json:"error"`
+}
+
+type DeviceFlowTokenGetter struct {
+	MessagePrinter func(string)
+	Sleeper        func(time.Duration)
+	Issuer         string
+}
+
+func NewDeviceFlowTokenGetter(issuer string) *DeviceFlowTokenGetter {
+	return &DeviceFlowTokenGetter{
+		MessagePrinter: func(s string) { fmt.Println(s) },
+		Sleeper:        time.Sleep,
+		Issuer:         issuer,
+	}
+}
+
+func (d *DeviceFlowTokenGetter) deviceFlow(clientID string) (string, error) {
+	deviceUrl := d.Issuer + "/device/code"
```

Comment:
Both request endpoints are arbitrary in the spec. Google uses a discovery endpoint to advertise which to use: https://developers.google.com/identity/protocols/oauth2/openid-connect#discovery

Reply:
Nice catch. Fixed this one. The discovery stuff is too buried in the oauth2 library we use to be useful so I made it configurable.